### ISSUE TITLE
Suggestion: Consider using Result instead of optional response and error

### ISSUE THREAD

**Issue by SebastianOsinski** [mapbox/mapbox-directions-swift#414]
This is merely a suggestion. We could use Result for completion handlers instead of all optional arguments:

```swift
public typealias RouteCompletionHandler = Result<([Waypoint], [Route]), DirectionsError> -> Void

public typealias MatchCompletionHandler = Result<[Match], DirectionsError> -> Void
```

This still can be done before 1.0.0 release. The only thing I am not sure about is `[Waypoint]` in `RouteCompletionHandler`. Is it always returned when routes are returned? Or is it possible to get waypoints but not routes or vice-versa?

**Comment by 1ec5** [comment]
Thanks for the suggestion! This overlaps with some of the proposal in #391 to encapsulate Route/Match in a response object. That refactoring is underway in #406.

We’ve been going back and forth about whether the options object and error should be members of the response object or live alongside it in the completion handler. The optionality is unfortunate, given that a nontrivial response is mutually exclusive with an error. That might be a case for using an enumeration with associated values. 🤔

A generic response type would be an interesting idea. So far we’ve been thinking about a response protocol, which would enforce the presence of a few members we think would be common to any navigation-related Mapbox API. But a generic type might be a cleaner way to go about that.

Please have a look at #406. We’d welcome suggestions, though we also expect these types to require additional rounds of refactoring, such as for waypoints in #388.

> The only thing I am not sure about is `[Waypoint]` in `RouteCompletionHandler`. Is it always returned when routes are returned? Or is it possible to get waypoints but not routes or vice-versa?

As things stand, the response objects would fall back to the waypoints in the options object:

https://github.com/mapbox/mapbox-directions-swift/blob/fe77075932d469473237bb5346dfcc5909124886/Sources/MapboxDirections/RouteResponse.swift#L42-L44

However, certain errors will pass no waypoints to the completion handler, while others will pass in the response object’s waypoints:

https://github.com/mapbox/mapbox-directions-swift/blob/fe77075932d469473237bb5346dfcc5909124886/Sources/MapboxDirections/Directions.swift#L187-L189 https://github.com/mapbox/mapbox-directions-swift/blob/fe77075932d469473237bb5346dfcc5909124886/Sources/MapboxDirections/Directions.swift#L194-L196

**Comment by 1ec5** [comment]
Oh whoops, I didn’t realize you were referring to the Result type in the standard library – that’s a good idea.

**Comment by JThramer** [comment]
Yeah this is a fantastic idea, and we should implement it as a fast-follow.

**Comment by 1ec5** [comment]
Fixed in #418.
